com.unity.ml-agents/Runtime/InferenceBrain/ModelRunner.cs

```diff
@@ -130,7 +127,15 @@ public void PutObservations(AgentInfo info, List<ISensor> sensors, Action<AgentA
                 sensors = sensors
             });
 
-            m_ActionFuncs.Add(new AgentIdActionPair { action = action, agentId = info.episodeId });
+            m_ActionId.Add(info.episodeId);
+
+            m_LastActionsReceived[info.episodeId] = null;
```

Comment:
Memory reuse optimization part 1: only set to null here if the key didn't already exist.